src/widget/transfer_function.ts

```diff
@@ -0,0 +1,1625 @@
+      positionArray[positionIndex + 1] = outputValue;
+
+      // Don't create a line for the last point
+      if (i === controlPoints.length - 1) break;
+      const lineBetweenPoints = vec4.fromValues(
+        inputValue,
+        outputValue,
+        normalizedControlPoints[i + 1].input,
+        normalizedControlPoints[i + 1].output,
+      );
+      positionArrayIndex = addLine(
+        linePositionArray,
+        positionArrayIndex,
+        lineBetweenPoints,
+      );
+    }
+    // Draw a horizontal line out from the last point
+    if (lineToRightEdge !== null) {
+      addLine(linePositionArray, positionArrayIndex, lineToRightEdge);
+    }
+
+    // Update buffers
+    this.controlPointsColorArray = colorArray;
+    this.controlPointsPositionArray = positionArray;
+    this.linePositionArray = linePositionArray;
+    this.controlPointsVertexBuffer.setData(this.controlPointsPositionArray);
+    this.controlPointsColorBuffer.setData(this.controlPointsColorArray);
+    this.linePositionBuffer.setData(this.linePositionArray);
+  }
+
+  private transferFunctionLineShader = this.registerDisposer(
+    (() => {
+      const builder = new ShaderBuilder(this.gl);
+      defineLineShader(builder);
+      builder.addAttribute("vec4", "aLineStartEnd");
+      builder.addOutputBuffer("vec4", "out_color", 0);
+      builder.addVarying("float", "vColor");
+      builder.setVertexMain(`
+vec4 start = vec4(aLineStartEnd[0], aLineStartEnd[1], 0.0, 1.0);
+vec4 end = vec4(aLineStartEnd[2], aLineStartEnd[3], 0.0, 1.0);
+emitLine(start, end, 1.0);
+`);
+      builder.setFragmentMain(`
+out_color = vec4(0.0, 1.0, 1.0, getLineAlpha());
+`);
+      return builder.build();
+    })(),
+  );
+
+  private transferFunctionShader = this.registerDisposer(
+    (() => {
+      const builder = new ShaderBuilder(this.gl);
+      builder.addAttribute("vec2", "aVertexPosition");
+      builder.addVarying("vec2", "vTexCoord");
+      builder.addOutputBuffer("vec4", "out_color", 0);
+      builder.addTextureSampler(
+        "sampler2D",
+        "uSampler",
+        transferFunctionSamplerTextureUnit,
+      );
+      builder.addUniform("float", "uTransferFunctionEnd");
+      builder.setVertexMain(`
+gl_Position = vec4(aVertexPosition, 0.0, 1.0);
+vTexCoord = (aVertexPosition + 1.0) / 2.0;
+`);
+      builder.setFragmentMain(`
+ivec2 texel = ivec2(floor(vTexCoord.x * uTransferFunctionEnd), 0);
+out_color = texelFetch(uSampler, texel, 0);
+`);
+      return builder.build();
+    })(),
+  );
+
+  private controlPointsShader = this.registerDisposer(
+    (() => {
+      const builder = new ShaderBuilder(this.gl);
+      builder.addAttribute("vec2", "aVertexPosition");
+      builder.addAttribute("vec3", "aVertexColor");
+      builder.addVarying("vec3", "vColor");
+      builder.addOutputBuffer("vec4", "out_color", 0);
+      builder.setVertexMain(`
+gl_Position = vec4(aVertexPosition, 0.0, 1.0);
+gl_PointSize = 14.0;
+vColor = aVertexColor;
+`);
```

Comment:
Comment that code below is to make the point circular with a visible border. The border color is black if the point is closer to white, and white is the point is closer to black.